lib/commands/web.js

```diff
@@ -255,21 +257,82 @@ extensions.translateWebCoords = async function translateWebCoords (coords) {
   }
 };
 
-extensions.checkForAlert = async function checkForAlert () { // eslint-disable-line require-await
-  return false;
+extensions.checkForAlert = async function checkForAlert (isRetroactive) {
+  if (!this.settings.getSettings().checkForModalDialogs) {
+    return;
+  }
+  let alert_ = null;
+  try {
+    // We need to check for native and js alerts (Location permission is a native alert!)
+    alert_ = await this.getAlertText();
+  } catch (err) {
+    // No alert found
+  }
+
+  if (alert_) {
+    let message;
+    if (isRetroactive) {
+      message = 'Alert detected after action had already started.' +
+        ' The action may have an unexpected side effect.';
+    } else {
+      message = 'Alert detected before action performed';
+    }
+    throw new errors.UnexpectedAlertOpenError(message);
+  }
+};
+
+extensions.executeAtom = async function executeAtom (atom, args, alwaysDefaultFrame = false) {
+  await this.checkForAlert();
+  let frames = alwaysDefaultFrame === true ? [] : this.curWebFrames;
+  let promise = this.remote.executeAtom(atom, args, frames);
+  return this.waitForAtom(promise);
+};
+
+extensions.executeAtomAsync = async function executeAtomAsync (atom, args, responseUrl) {
+  await this.checkForAlert();
+  // save the resolve and reject methods of the promise to be waited for
+  let promise = new B((resolve, reject) => {
+    this.asyncPromise = {resolve, reject};
+  });
+  await this.remote.executeAtomAsync(atom, args, this.curWebFrames, responseUrl);
+  return await this.waitForAtom(promise);
 };
 
 extensions.waitForAtom = async function waitForAtom (promise) {
-  const started = process.hrtime();
-  try {
-    return this.parseExecuteResponse(await B.resolve(promise)
-      .timeout(ATOM_WAIT_TIMEOUT));
-  } catch (err) {
-    if (err instanceof B.TimeoutError) {
-      throw new Error(`Did not get any response after ${process.hrtime(started)[0]}s`);
+  let res = null;
+  let timedOut = false;
+  let done = false;
+  let error = null;
+  // Do the action asynchronously, so we can monitor it for taking too long.
+  promise.then((res) => {
+    done = true;
+    return res;
+  })
+  .catch((err) => {
+    log.debug(`Error received while executing atom: ${err.message}`);
+    // error gets swallowed, so save and check later
+    error = err;
+  });
+  // Wait until either action succeeds or it's taking too long and we want to check for alerts.
+  let timeoutPromise = B.delay(1000).then(() => timedOut = true);
+  // eslint-disable-next-line promise/no-native
+  await Promise.race([promise, timeoutPromise]);
```

Comment:
Use [bluebird#race](http://bluebirdjs.com/docs/api/promise.race.html).

Reply:
Bleubird promises also provide the native `timeout` method